### DISCUSSION TITLE
Shift Origin

### DISCUSSION THREAD

**Discussion by JSandusky** [bulletphysics/bullet3#3947] [Q&A] [answered]
Objective is the same as PhysX shift-origin, context is also NOT robotics.

The objective is to translate everything by vector without reconstructing anything or shifting each object individually and resolving the consequences of the shift for each object's shift individually. Instead translating the entirety of the simulation without changing it's tree. The wake/sleep states of all shifted objects should remain the same as they were before the shift.

From what I can tell (when using Dbvt) there should be nothing wrong with walking the tree and shifting the min/max of the nodes and then doing the appropriate offsets for btSoftBody and btCollisionObject. Everything else (contacts, joints, etc) appears to be in one local space or another.

Is this a safe assumption?

**Comment by JSandusky** [comment, marked as answer]
This appears to hold true at least for a trivial scene of a stack of boxes and ragdolls. Constraints and contacts all appear correct through shifts.

```cpp
void btDbvtNode::offset(const btVector3& ofs)
{
    volume.offset(ofs);
    if (isinternal())
    {
        childs[0]->offset(ofs);
        childs[1]->offset(ofs);
    }
    else
    {
        if (auto prox = (btDbvtProxy*)data)
        {
            if (auto obj = ((btCollisionObject*)prox->m_clientObject))
            {
                if (obj->getInternalType() == btCollisionObject::CO_SOFT_BODY)
                {
                    auto softBody = ((btSoftBody*)obj);
                    auto transform = softBody->getWorldTransform();
                    auto origin = transform.getOrigin() + ofs;
                    transform.setOrigin(origin);
                    obj->setWorldTransform(transform);
                }
                else if (obj->getInternalType() == btCollisionObject::CO_RIGID_BODY)
                {
                    auto transform = obj->getWorldTransform();
                    auto origin = transform.getOrigin() + ofs;
                    transform.setOrigin(origin);
                    obj->setWorldTransform(transform);
                }
            }
        }
    }
}
```

**Comment by BluePrintRandom** [comment]
recently I did the inverse of this

I actually shift a terrain mesh periodically on a grid -> with the
terrain marked dynamic, and set it's linV to 0 and all the objects sit in
place

sort of like pulling a tablecloth out from a bunch of stuff and it staying
https://www.youtube.com/watch?v=UPC2uJVLSh0
